# Assignment SVM Salary Day

## Prepare a classification model using SVM for salary data 

__Data Description__

- age -- age of a person
- workclass	-- A work class is a grouping of work 
- education	-- Education of an individuals	
- maritalstatus -- Marital status of an individulas	
- occupation	 -- occupation of an individuals
- relationship -- 	
- race --  Race of an Individual
- sex --  Gender of an Individual
- capitalgain --  profit received from the sale of an investment	
- capitalloss	-- A decrease in the value of a capital asset
- hoursperweek -- number of hours work per week	
- native -- Native of an individual
- Salary -- salary of an individual

In [1]:
# SVM Classification
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn import svm
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.model_selection import train_test_split, cross_val_score

In [2]:
# loading salary training data 
salary_data=pd.read_csv("SalaryData_Train(1).csv")

In [3]:
salary_data.head()

age          workclass   education  educationno        maritalstatus  \
0   39          State-gov   Bachelors           13        Never-married   
1   50   Self-emp-not-inc   Bachelors           13   Married-civ-spouse   
2   38            Private     HS-grad            9             Divorced   
3   53            Private        11th            7   Married-civ-spouse   
4   28            Private   Bachelors           13   Married-civ-spouse   

           occupation    relationship    race      sex  capitalgain  \
0        Adm-clerical   Not-in-family   White     Male         2174   
1     Exec-managerial         Husband   White     Male            0   
2   Handlers-cleaners   Not-in-family   White     Male            0   
3   Handlers-cleaners         Husband   Black     Male            0   
4      Prof-specialty            Wife   Black   Female            0   

   capitalloss  hoursperweek          native  Salary  
0            0            40   United-States   <=50K  
1            0            13   United-States   <=50K  
2            0            40   United-States   <=50K  
3            0            40   United-States   <=50K  
4            0            40            Cuba   <=50K

In [4]:
salary_data.shape

(30161, 14)

In [5]:
salary_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30161 entries, 0 to 30160
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   age            30161 non-null  int64 
 1   workclass      30161 non-null  object
 2   education      30161 non-null  object
 3   educationno    30161 non-null  int64 
 4   maritalstatus  30161 non-null  object
 5   occupation     30161 non-null  object
 6   relationship   30161 non-null  object
 7   race           30161 non-null  object
 8   sex            30161 non-null  object
 9   capitalgain    30161 non-null  int64 
 10  capitalloss    30161 non-null  int64 
 11  hoursperweek   30161 non-null  int64 
 12  native         30161 non-null  object
 13  Salary         30161 non-null  object
dtypes: int64(5), object(9)
memory usage: 3.2+ MB


##### inspecting object datatypes columns

 1   workclass      30161 non-null  object<br>
 2   education      30161 non-null  object<br>
 4   maritalstatus  30161 non-null  object<br>
 5   occupation     30161 non-null  object<br>
 6   relationship   30161 non-null  object<br>
 7   race           30161 non-null  object<br>
 8   sex            30161 non-null  object<br>
 12  native         30161 non-null  object<br>
 13  Salary         30161 non-null  object
 


In [6]:
salary_data.education.unique()

array([' Bachelors', ' HS-grad', ' 11th', ' Masters', ' 9th',
       ' Some-college', ' Assoc-acdm', ' 7th-8th', ' Doctorate',
       ' Assoc-voc', ' Prof-school', ' 5th-6th', ' 10th', ' Preschool',
       ' 12th', ' 1st-4th'], dtype=object)

In [7]:
def unique_value(df, cols_to_include):
    for v in cols_to_include:
        x=df[cols_to_include]
        #print ("Unique vaues for: {} is: {}".format(df.v.unique()))
        print("Unique value for: {} are:".format(v) )
        print(df[v].unique(),"\n")
        

In [8]:
unique_value(df=salary_data, cols_to_include=['workclass','education','maritalstatus','occupation','relationship','race','sex','native','Salary'])

Unique value for: workclass are:
[' State-gov' ' Self-emp-not-inc' ' Private' ' Federal-gov' ' Local-gov'
 ' Self-emp-inc' ' Without-pay'] 

Unique value for: education are:
[' Bachelors' ' HS-grad' ' 11th' ' Masters' ' 9th' ' Some-college'
 ' Assoc-acdm' ' 7th-8th' ' Doctorate' ' Assoc-voc' ' Prof-school'
 ' 5th-6th' ' 10th' ' Preschool' ' 12th' ' 1st-4th'] 

Unique value for: maritalstatus are:
[' Never-married' ' Married-civ-spouse' ' Divorced'
 ' Married-spouse-absent' ' Separated' ' Married-AF-spouse' ' Widowed'] 

Unique value for: occupation are:
[' Adm-clerical' ' Exec-managerial' ' Handlers-cleaners' ' Prof-specialty'
 ' Other-service' ' Sales' ' Transport-moving' ' Farming-fishing'
 ' Machine-op-inspct' ' Tech-support' ' Craft-repair' ' Protective-serv'
 ' Armed-Forces' ' Priv-house-serv'] 

Unique value for: relationship are:
[' Not-in-family' ' Husband' ' Wife' ' Own-child' ' Unmarried'
 ' Other-relative'] 

Unique value for: race are:
[' White' ' Black' ' Asian-Pac-Islande

In [9]:
cols_to_include=['workclass','education','maritalstatus','occupation','relationship','race','sex','native','Salary']

In [10]:
# enocde all categorical values to o and 1
#get all categorical columns
cat_columns = salary_data.select_dtypes(['object']).columns

#convert all categorical columns to numeric
salary_data[cat_columns] = salary_data[cat_columns].apply(lambda x: pd.factorize(x)[0])

In [11]:
salary_data

age  workclass  education  educationno  maritalstatus  occupation  \
0       39          0          0           13              0           0   
1       50          1          0           13              1           1   
2       38          2          1            9              2           2   
3       53          2          2            7              1           2   
4       28          2          0           13              1           3   
...    ...        ...        ...          ...            ...         ...   
30156   27          2          6           12              1           9   
30157   40          2          1            9              1           8   
30158   58          2          1            9              6           0   
30159   22          2          1            9              0           0   
30160   52          5          1            9              1           1   

       relationship  race  sex  capitalgain  capitalloss  hoursperweek  \
0                 0     0    0         2174            0            40   
1                 1     0    0            0            0            13   
2                 0     0    0            0            0            40   
3                 1     1    0            0            0            40   
4                 2     1    1            0            0            40   
...             ...   ...  ...          ...          ...           ...   
30156             2     0    1            0            0            38   
30157             1     0    0            0            0            40   
30158             4     0    1            0            0            40   
30159             3     0    0            0            0            20   
30160             2     0    1        15024            0            40   

       native  Salary  
0           0       0  
1           0       0  
2           0       0  
3           0       0  
4           1       0  
...       ...     ...  
30156       0       0  
30157       0       1  
30158       0       0  
30159       0       0  
30160       0       1  

[30161 rows x 14 columns]

In [19]:
X_train=salary_data.iloc[:,0:13]
y_train=salary_data.iloc[:,13]

In [20]:
X_train.shape

(30161, 13)

In [14]:
# X_train, X_test, y_train, y_test = train_test_split(X,Y, test_size = 0.3)

In [15]:
# X_train.shape, y_train.shape, X_test.shape, y_test.shape

### Grid Search CV

In [16]:
# clf = SVC()
# param_grid = [{'kernel':['rbf'],'gamma':[100,50,0.5,0.1,0.01,0.001,0.0001],
#                'C':[15,14,13,12,11,10,0.1,0.001,0.0001]}]
# gsv = GridSearchCV(clf,param_grid,cv=10)
# gsv.fit(X_train,y_train)

In [21]:
clf = SVC()
param_grid = [{'kernel':['rbf'],'gamma':[0.5,0.1,0.01,0.001,0.0001],
               'C':[10,0.1,0.001,0.0001]}]
gsv = GridSearchCV(clf,param_grid,cv=10)
gsv.fit(X_train,y_train)

GridSearchCV(cv=10, estimator=SVC(),
             param_grid=[{'C': [10, 0.1, 0.001, 0.0001],
                          'gamma': [0.5, 0.1, 0.01, 0.001, 0.0001],
                          'kernel': ['rbf']}])

In [22]:
gsv.best_params_ , gsv.best_score_ 

({'C': 10, 'gamma': 0.01, 'kernel': 'rbf'}, 0.8509336241404807)

In [25]:
clf = SVC(C= 10, gamma = 0.01)
clf.fit(X_train , y_train)
y_pred_train= clf.predict(X_train)
acc = accuracy_score(y_train, y_pred_train) * 100
print("Accuracy =", acc)
print(confusion_matrix(y_train, y_pred_train))

Accuracy = 88.12373595039952
[[21856   797]
 [ 2785  4723]]


## loading SalaryData_Test(1).csv for running model on test data

In [29]:
salary_test_data=pd.read_csv("SalaryData_Test(1).csv")

In [30]:
salary_test_data.head()

age   workclass      education  educationno        maritalstatus  \
0   25     Private           11th            7        Never-married   
1   38     Private        HS-grad            9   Married-civ-spouse   
2   28   Local-gov     Assoc-acdm           12   Married-civ-spouse   
3   44     Private   Some-college           10   Married-civ-spouse   
4   34     Private           10th            6        Never-married   

           occupation    relationship    race    sex  capitalgain  \
0   Machine-op-inspct       Own-child   Black   Male            0   
1     Farming-fishing         Husband   White   Male            0   
2     Protective-serv         Husband   White   Male            0   
3   Machine-op-inspct         Husband   Black   Male         7688   
4       Other-service   Not-in-family   White   Male            0   

   capitalloss  hoursperweek          native  Salary  
0            0            40   United-States   <=50K  
1            0            50   United-States   <=50K  
2            0            40   United-States    >50K  
3            0            40   United-States    >50K  
4            0            30   United-States   <=50K

In [31]:
salary_test_data.shape

(15060, 14)

In [32]:
# enocde all categorical values to o and 1
#get all categorical columns
cat_columns = salary_test_data.select_dtypes(['object']).columns

#convert all categorical columns to numeric
salary_test_data[cat_columns] = salary_test_data[cat_columns].apply(lambda x: pd.factorize(x)[0])

In [33]:
salary_test_data.head()

age  workclass  education  educationno  maritalstatus  occupation  \
0   25          0          0            7              0           0   
1   38          0          1            9              1           1   
2   28          1          2           12              1           2   
3   44          0          3           10              1           0   
4   34          0          4            6              0           3   

   relationship  race  sex  capitalgain  capitalloss  hoursperweek  native  \
0             0     0    0            0            0            40       0   
1             1     1    0            0            0            50       0   
2             1     1    0            0            0            40       0   
3             1     0    0         7688            0            40       0   
4             2     1    0            0            0            30       0   

   Salary  
0       0  
1       0  
2       1  
3       1  
4       0

In [34]:
X_test=salary_test_data.iloc[:,0:13]
y_test=salary_test_data.iloc[:,13]

In [36]:
y_pred_test2 = clf.predict(X_test)
acc = accuracy_score(y_test, y_pred_test2) * 100
print("Accuracy =", acc)
print(confusion_matrix(y_test, y_pred_test2))

Accuracy = 82.60292164674635
[[10951   409]
 [ 2211  1489]]
